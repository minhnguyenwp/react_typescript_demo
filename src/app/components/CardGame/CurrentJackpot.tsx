import React, { useEffect, useState } from 'react';
import styled from 'styled-components/macro';

import { useRetrieveJackPots } from 'services/jackpots';
import { COLORS } from 'styles/global-colors';
import { getAmountJackPot } from 'helpers';

interface CurrentJackpotProps {
  gameId: string;
}

export const CurrentJackpot: React.FC<CurrentJackpotProps> = React.memo(
  props => {
    const { gameId } = props;
    const [amount, setAmount] = useState<number>(0);
    const { data: response } = useRetrieveJackPots(
      {},
      {
        cacheTime: 0,
        refetchOnWindowFocus: false,
        refetchInterval: 3000,
      },
    );
    const { data } = response || {};

    useEffect(() => {
      const currentAmount = getAmountJackPot(gameId, data) || 0;
      setAmount(prev => prev + currentAmount);
    }, [data, gameId]);

    return (
      <>
        {amount > 0 && (
          <Wrapper>
            <span className="amount-jackpot">${amount}</span>
          </Wrapper>
        )}
      </>
    );
  },
  (prevProps, nextProps) => prevProps.gameId === nextProps.gameId,
);

const Wrapper = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.5);

  .amount-jackpot {
    color: ${COLORS.grey20};
    font-size: 11px;
    padding: 5px 10px;
    line-height: 1;
  }
`;
