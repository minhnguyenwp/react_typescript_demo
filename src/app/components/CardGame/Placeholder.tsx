import React from 'react';
import Skeleton from 'react-loading-skeleton';

const Placeholder: React.FC = React.memo(() => {
  return (
    <div className="row">
      {Array.from({ length: 25 }, (_, k) => (
        <div key={k} className="cell-20 game-item">
          <Skeleton width="224px" height="134px" borderRadius="10px" />
        </div>
      ))}
    </div>
  );
});

export default Placeholder;
