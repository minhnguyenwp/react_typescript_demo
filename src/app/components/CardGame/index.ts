export { default as CardGame } from './CardGame';
export { default as CardGamePlaceholder } from './Placeholder';
