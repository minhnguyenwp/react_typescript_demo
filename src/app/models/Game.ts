export interface Game {
  categories: string[];
  name: string;
  image: string;
  id: string;
}
