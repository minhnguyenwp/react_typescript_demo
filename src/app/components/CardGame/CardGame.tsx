import React, { memo } from 'react';
import { useLocation } from 'react-router';
import qs from 'query-string';
import styled from 'styled-components/macro';
import Img from 'react-cool-img';

import { RIBBON_NEW_VALUE, RIBBON_TOP_VALUE } from 'app/constants';
import { Game } from 'app/models/Game';
import { getRibbonByCate } from 'helpers';
import { COLORS } from 'styles/global-colors';
import { imageLoading, ribbonNew, ribbonTop } from 'utils/assetUrl';

// Components
import { CurrentJackpot } from './CurrentJackpot';
interface Props {
  game: Game;
}

const CardGame: React.FC<Props> = props => {
  const { game } = props;
  const location = useLocation();
  const query = qs.parse(location?.search);
  const currentCate = query?.categories;
  const { id, image, name } = game || {};

  const renderRibbonImage = (game: Game): JSX.Element | null => {
    const currentRibbon = getRibbonByCate(game, currentCate);
    if (currentRibbon === RIBBON_NEW_VALUE) {
      return <Ribbon className="ribbon" imgUrl={ribbonNew} />;
    }
    if (currentRibbon === RIBBON_TOP_VALUE) {
      return <Ribbon className="ribbon" imgUrl={ribbonTop} />;
    }
    return null;
  };

  return (
    <div className="cell-20 game-item">
      <Card key={id} className="card">
        <Img
          placeholder={imageLoading}
          src={image}
          alt={name}
          className="game-pic"
        />
        {renderRibbonImage(game)}
        <CurrentJackpot gameId={id} />
        <div className="display-card-group m-aligner">
          <div>
            <p className="g-name">{name}</p>
            <ButtonPlay>Play</ButtonPlay>
          </div>
        </div>
      </Card>
    </div>
  );
};

const MemoizedCardGame = memo(
  CardGame,
  (prevProps, nextProps) =>
    JSON.stringify(prevProps.game) === JSON.stringify(nextProps.game),
);

export default MemoizedCardGame;

const ButtonPlay = styled.span`
  background: linear-gradient(
    266.53deg,
    ${COLORS.seaSerpent} 0%,
    ${COLORS.accentGreen} 100%
  );
  border-radius: 6px;
  color: ${COLORS.white};
  padding: 5px 15px;
  font-size: 14px;
  line-height: 1;
  outline: none;
  border: none;
  display: inline-block;
`;

const Ribbon = styled.span<{ imgUrl: string }>`
  position: absolute;
  right: 0;
  top: 0;
  width: 60px;
  height: 60px;
  background: ${props =>
    `url(${props?.imgUrl && props?.imgUrl}) no-repeat top right`};
  background-size: 60px;
`;

const Card = styled.a`
  display: block;
  transition: transform 0.25s ease-out;
  position: relative;
  border-radius: 10px;
  overflow: hidden;
  cursor: pointer;
  width: 100%;
  height: 100%;
  max-height: 132px;
  &:hover {
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
    z-index: 2;
    .display-card-group {
      opacity: 1;
      pointer-events: auto;
      z-index: 3;
    }
    @media (max-width: 1024px) {
      -webkit-transform: scale(1);
      -ms-transform: scale(1);
      transform: scale(1);
      .display-card-group {
        opacity: 0;
        pointer-events: none;
        z-index: -1;
      }
    }
  }
  img.game-pic {
    width: 100%;
    height: auto;
    z-index: 1;
  }
  .display-card-group {
    position: absolute;
    top: 0;
    right: 0;
    left: 0;
    bottom: 0;
    z-index: 4;
    background-color: rgba(0, 0, 0, 0.5);
    text-align: center;

    transition: all 0.25s ease-out;
    opacity: 0;
    pointer-events: none;
    z-index: -1;

    .g-name {
      color: ${COLORS.white};
      margin: 0;
      margin-bottom: 10px;
      text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.6);
    }
  }
`;
