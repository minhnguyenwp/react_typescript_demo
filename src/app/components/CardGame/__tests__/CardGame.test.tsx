import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { render } from '@testing-library/react';

import CardGame from '../CardGame';
import { Game } from 'app/models/Game';

const renderer = createRenderer();

const game: Game = {
  categories: ['top', 'slots', 'new'],
  id: 'Test',
  image: '//stage.whgstage.com/scontent/images/games/NETHEWISHMASTER.jpg',
  name: 'Test Game',
};

describe('CardGame', () => {
  renderer.render(<CardGame game={game} />);
  const renderedOutput = renderer.getRenderOutput();

  it('Should render the correct snapshot', () => {
    expect(renderedOutput).toMatchSnapshot();
  });

  it('Should render game name', () => {
    const gameTag = render(<CardGame game={game} />);
    expect(gameTag.container.querySelector('p')?.innerHTML).toBe(game.name);
  });
});
