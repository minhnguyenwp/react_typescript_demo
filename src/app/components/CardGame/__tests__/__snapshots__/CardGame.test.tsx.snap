// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGame Should render the correct snapshot 1`] = `
<div
  className="cell-20 game-item"
>
  <CardGame__Card
    className="card"
  >
    <Memo
      alt="Test Game"
      className="game-pic"
      placeholder="loading.gif"
      src="//stage.whgstage.com/scontent/images/games/NETHEWISHMASTER.jpg"
    />
    <CardGame__Ribbon
      className="ribbon"
      imgUrl="new_ribbon.png"
    />
    <Memo
      gameId="Test"
    />
    <div
      className="display-card-group m-aligner"
    >
      <div>
        <p
          className="g-name"
        >
          Test Game
        </p>
        <CardGame__ButtonPlay>
          Play
        </CardGame__ButtonPlay>
      </div>
    </div>
  </CardGame__Card>
</div>
`;
